Sources/NIOSOCKS/Messages/ServerResponse.swift

```diff
@@ -0,0 +1,112 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+// MARK: - ServerResponse
+
+/// The SOCKS Server's response to the client's request
+/// indicating if the request succeeded or failed.
+struct ServerResponse: Hashable {
+    
+    /// The SOCKS protocol version - we currently only support v5.
+    public let version: UInt8 = 5
+    
+    /// The status of the connection - used to check if the request
+    /// succeeded or failed.
+    public var reply: Reply
+    
+    /// The host address.
+    public var boundAddress: AddressType
+    
+    /// Creates a new `ServerResponse`.
+    /// - parameter reply: The status of the connection - used to check if the request
+    /// succeeded or failed.
+    /// - parameter boundAddress: The host address.
+    public init(reply: Reply, boundAddress: AddressType) {
+        self.reply = reply
+        self.boundAddress = boundAddress
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readServerResponse() throws -> ServerResponse? {
+        return try self.parseUnwindingIfNeeded { buffer in
+            guard
+                let version = buffer.readInteger(as: UInt8.self),
+                let reply = buffer.readInteger(as: UInt8.self).map({ Reply(value: $0) }),
+                let reserved = buffer.readInteger(as: UInt8.self),
+                let boundAddress = try buffer.readAddressType()
+            else {
+                throw MissingBytes()
+            }
+            
+            guard reserved == 0x0 else {
```

Comment:
It would be better to do these checks before we attempt to read the address or other fields. That way we'll fail early, rather than block waiting for bytes that may never come because the message isn't the type we expected.